pkg/discovery/systemd.go

```diff
@@ -50,111 +45,65 @@ func (c *SystemdConfig) NewDiscoverer(d DiscovererOptions) (Discoverer, error) {
-	ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
-	defer cancel()
-
-	cmd := exec.CommandContext(ctx, "systemctl", "list-units", "--type=service", "--state=running")
-	stdout, err := cmd.StdoutPipe()
+	conn, err := systemd.NewWithContext(ctx)
 	if err != nil {
-		return nil, err
-	}
-	if err := cmd.Start(); err != nil {
-		return nil, err
+		level.Error(c.logger).Log("msg", "failed to connect to systemd D-Bus API", "err", err)
+		return err
 	}
+	defer conn.Close()
 
-	units := parseUnitList(stdout)
-	if err := cmd.Wait(); err != nil {
-		return nil, err
+	isSubStateChanged := func(u1, u2 *systemd.UnitStatus) bool {
+		return u1.SubState != u2.SubState
 	}
-	return units, nil
-}
 
-func parseUnitList(stdout io.Reader) []string {
-	units := []string{}
-	s := bufio.NewScanner(stdout)
-	for s.Scan() {
-		line := strings.TrimSpace(s.Text())
-		if strings.HasPrefix(line, "UNIT") ||
-			strings.HasPrefix(line, "LOAD") ||
-			strings.HasPrefix(line, "ACTIVE") ||
-			strings.HasPrefix(line, "SUB") {
-			continue
-		}
-		fields := strings.Fields(line)
-		if len(fields) < 1 {
-			continue
-		}
-		units = append(units, fields[0])
-	}
-	if len(units) > 0 {
-		units = units[:len(units)-1]
+	isNotService := func(name string) bool {
+		return !strings.HasSuffix(name, ".service")
 	}
-	return units
-}
 
-func (c *SystemdDiscoverer) pids(ctx context.Context, unit string) ([]int, error) {
-	ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
-	defer cancel()
+	updateCh, errCh := conn.SubscribeUnitsCustom(time.Second, 0, isSubStateChanged, isNotService)
 
-	cmd := exec.CommandContext(ctx, "systemctl", "show", "--property", "MainPID", "--value", unit)
-	output, err := cmd.Output()
-	if err != nil {
-		return nil, err
-	}
+	for {
+		select {
+		case update := <-updateCh:
+			var groups []*Group
+
+			for unit, status := range update {
+				if status == nil || status.SubState != "running" {
+					groups = append(groups, &Group{Source: unit})
+					continue
+				}
+
+				mainPIDProperty, err := conn.GetServicePropertyContext(ctx, unit, "MainPID")
+				if err != nil {
+					level.Error(c.logger).Log("msg", "failed to get MainPID property for service", "err", err, "unit", unit)
+					continue
+				}
+
+				pid, ok := mainPIDProperty.Value.Value().(uint32)
+				if !ok {
+					level.Error(c.logger).Log("msg", "failed to assert type of PID", "unit", unit)
+					continue
+				}
+
+				groups = append(groups, &Group{
+					Targets: []model.LabelSet{{}},
+					Labels: model.LabelSet{
+						model.LabelName("systemd_unit"): model.LabelValue(unit),
+					},
+					Source: unit,
+					PIDs:   []int{int(pid)},
+				})
+			}
 
-	pid, err := strconv.Atoi(strings.TrimSpace(string(output)))
-	if err != nil {
-		return nil, err
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case up <- groups:
+			}
+		case err := <-errCh:
+			level.Error(c.logger).Log("msg", "received error from systemd D-Bus API", "err", err)
```

Comment:
```suggestion
			level.Warn(c.logger).Log("msg", "received error from systemd D-Bus API", "err", err)
```

nit: Let's lower the level. It's not an error that stops operations.

Reply:
Same goes other error levels.